src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSourceProviders/WorkspaceDocumentsAndProjectDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            var fullSolutionAnalysisEnabled = globalOptions.IsFullSolutionAnalysisEnabled(project.Language, out var compilerFullSolutionAnalysisEnabled, out var analyzersFullSolutionAnalysisEnabled);
+            if (!fullSolutionAnalysisEnabled && !codeAnalysisService.HasProjectBeenAnalyzed(project.Id))
+                return;
+
+            Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer = !compilerFullSolutionAnalysisEnabled || !analyzersFullSolutionAnalysisEnabled
+                ? ShouldIncludeAnalyzer : null;
+
+            AddDocumentSources(project.Documents);
+            AddDocumentSources(project.AdditionalDocuments);
+
+            // If all features are enabled for source generated documents, then compute todo-comments/diagnostics for them.
+            if (enableDiagnosticsInSourceGeneratedFiles)
+            {
+                var sourceGeneratedDocuments = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
+                AddDocumentSources(sourceGeneratedDocuments);
+            }
+
+            // Finally, add the appropriate FSA or CodeAnalysis project source to get project specific diagnostics, not associated with any document.
+            AddProjectSource();
+
+            return;
+
+            void AddDocumentSources(IEnumerable<TextDocument> documents)
+            {
+                foreach (var document in documents)
+                {
+                    if (!WorkspaceDiagnosticSourceHelpers.ShouldSkipDocument(context, document))
+                    {
+                        // Add the appropriate FSA or CodeAnalysis document source to get document diagnostics.
+                        var documentDiagnosticSource = fullSolutionAnalysisEnabled
+                            ? AbstractWorkspaceDocumentDiagnosticSource.CreateForFullSolutionAnalysisDiagnostics(document, diagnosticAnalyzerService, shouldIncludeAnalyzer)
+                            : AbstractWorkspaceDocumentDiagnosticSource.CreateForCodeAnalysisDiagnostics(document, codeAnalysisService);
+                        result.Add(documentDiagnosticSource);
+                    }
+                }
+            }
+
+            void AddProjectSource()
+            {
+                var projectDiagnosticSource = fullSolutionAnalysisEnabled
+                    ? AbstractProjectDiagnosticSource.CreateForFullSolutionAnalysisDiagnostics(project, diagnosticAnalyzerService, shouldIncludeAnalyzer)
+                    : AbstractProjectDiagnosticSource.CreateForCodeAnalysisDiagnostics(project, codeAnalysisService);
+                result.Add(projectDiagnosticSource);
+            }
+
+            bool ShouldIncludeAnalyzer(DiagnosticAnalyzer analyzer)
+                => analyzer.IsCompilerAnalyzer() ? compilerFullSolutionAnalysisEnabled : analyzersFullSolutionAnalysisEnabled;
+        }
```

Comment:
i assume this was copied without changes?

Reply:
correct